File: apps/cli/src/commands/hints.rs
use crate::cli::{Cli, HintsArgs};
use crate::commands::{build, gas, write};
use crate::error::AppResult;
use crate::output::{self, Meta};
use serde::Serialize;
use std::fs;

#[derive(Debug, Serialize)]
struct HintsData {
    target: String,
    file: String,
    contract: String,
    project_root: String,
    diagnostics: Vec<build::Diagnostic>,
    gas_hints: Vec<GasHint>,
}

#[derive(Debug, Serialize)]
struct GasHint {
    signature: String,
    gas: String,
    finite: bool,
    signal: write::GasSignal,
    line: Option<usize>,
    message: String,
}

pub fn run(cli: &Cli, args: &HintsArgs) -> AppResult<()> {
    let target = target_from_args(args);
    let build_data = build::build_data(cli, Some(&target))?;
    let gas_data = gas::compile_data(cli, &target)?;
    let source = fs::read_to_string(&args.file).unwrap_or_default();
    let gas_hints = gas_data
        .functions
        .iter()
        .map(|function| gas_hint(function, &source))
        .collect::<Vec<_>>();
    let data = HintsData {
        target,
        file: args.file.display().to_string(),
        contract: gas_data.contract,
        project_root: gas_data.project_root,
        diagnostics: build_data.diagnostics,
        gas_hints,
    };

    if cli.json {
        let mut meta = Meta::new("hints");
        meta.project_root = Some(data.project_root.clone());
        output::print_json(data, meta)
    } else {
        println!("Hints: {}", data.target);
        println!("  diagnostics: {}", data.diagnostics.len());
        println!("  gas hints: {}", data.gas_hints.len());
        Ok(())
    }
}

fn target_from_args(args: &HintsArgs) -> String {
    let file = args.file.display().to_string();
    args.contract
        .as_ref()
        .map_or(file.clone(), |contract| format!("{file}:{contract}"))
}

fn gas_hint(function: &gas::FunctionGas, source: &str) -> GasHint {
    let line = function_line(source, &function.signature);
    let gas_label = if function.finite {
        function.gas.clone()
    } else {
        "infinite".to_string()
    };
    GasHint {
        signature: function.signature.clone(),
        gas: function.gas.clone(),
        finite: function.finite,
        signal: function.signal.clone(),
        line,
        message: format!("gas: {gas_label}"),
    }
}

fn function_line(source: &str, signature: &str) -> Option<usize> {
    let name = signature.split_once('(')?.0;
    let needle = format!("function {name}");
    let public_var = format!(" public {name}");
    source
        .lines()
        .position(|line| {
            line.contains(&needle) || (line.contains(&public_var) && line.trim_end().ends_with(';'))
        })
        .map(|index| index + 1)
}

File: packages/cli/src/commands/hints.ts
import { createSuccessEnvelope } from "@consol/protocol";
import { runForgeBuild } from "@consol/foundry";
import { readFileSync } from "node:fs";
import { isAbsolute, resolve } from "node:path";
import type { GlobalArgs } from "../args";
import type { CliEnv, CliResult } from "../main";
import { VERSION } from "../version";
import { parseBuildDiagnostics } from "./diagnostics";
import { createGasCompileData, type FunctionGas } from "./gas";

export type RunHintsCommandInput = {
  readonly globals: GlobalArgs;
  readonly commandArgs: readonly string[];
  readonly cwd: string;
  readonly env: CliEnv;
};

export async function runHintsCommand(input: RunHintsCommandInput): Promise<CliResult> {
  const file = flagValue(input.commandArgs, "--file") ?? "";
  const contract = flagValue(input.commandArgs, "--contract");
  const target = contract === undefined ? file : `${file}:${contract}`;
  const gas = createGasCompileData({
    cwd: input.cwd,
    target,
    ...(input.globals.project === undefined ? {} : { projectRoot: input.globals.project }),
  });
  const build = await runForgeBuild({
    cwd: gas.project_root,
    projectRoot: gas.project_root,
    env: input.env,
  });
  const source = readFileSync(sourceReadPath(input.cwd, file), "utf8");
  const data = {
    target,
    file,
    contract: gas.contract,
    project_root: gas.project_root,
    diagnostics: parseBuildDiagnostics(build.stdout, build.stderr),
    gas_hints: gas.functions.map((functionGas) => gasHint(functionGas, source)),
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "hints",
        project_root: data.project_root,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  return { exitCode: 0, stdout: `Hints: ${data.target}\n`, stderr: "" };
}

function gasHint(functionGas: FunctionGas, source: string): unknown {
  const gasLabel = functionGas.finite ? functionGas.gas : "infinite";
  return {
    signature: functionGas.signature,
    gas: functionGas.gas,
    finite: functionGas.finite,
    signal: functionGas.signal,
    line: functionLine(source, functionGas.signature),
    message: `gas: ${gasLabel}`,
  };
}

function functionLine(source: string, signature: string): number | null {
  const name = signature.split("(")[0];
  if (name === undefined || name === "") {
    return null;
  }

  const needle = `function ${name}`;
  const publicVar = ` public ${name}`;
  const index = source.split(/\r?\n/).findIndex((line) => {
    return line.includes(needle) || (line.includes(publicVar) && line.trimEnd().endsWith(";"));
  });
  return index === -1 ? null : index + 1;
}

function flagValue(commandArgs: readonly string[], flag: string): string | undefined {
  const index = commandArgs.indexOf(flag);
  return index === -1 ? undefined : commandArgs[index + 1];
}

function sourceReadPath(cwd: string, file: string): string {
  return isAbsolute(file) ? file : resolve(cwd, file);
}
